worker/caasapplicationprovisioner/application.go

```diff
@@ -96,20 +95,61 @@ func (a *appWorker) Wait() error {
 }
 
 func (a *appWorker) loop() error {
-	charmURL, err := a.facade.ApplicationCharmURL(a.name)
-	if errors.IsNotFound(err) {
-		a.logger.Debugf("application %q removed", a.name)
-		return nil
-	} else if err != nil {
-		return errors.Annotatef(err, "failed to get charm url for application")
-	}
-	charmInfo, err := a.facade.CharmInfo(charmURL.String())
+	shouldExit, err := a.verifyCharmUpgraded()
 	if err != nil {
-		return errors.Annotatef(err, "failed to get application charm deployment metadata for %q", a.name)
+		return errors.Trace(err)
 	}
-	if charmInfo == nil ||
-		charm.MetaFormat(charmInfo.Charm()) < charm.FormatV2 {
-		return errors.Errorf("charm version 2 or greater required")
+	if shouldExit {
+		return nil
+	}
+
+	// If the application has an operator pod due to an upgrade-charm from a
+	// pod-spec charm to a sidecar charm, delete it. Also delete workload pod.
+	const maxDeleteLoops = 20
+	for i := 0; ; i++ {
+		if i >= maxDeleteLoops {
+			return fmt.Errorf("couldn't delete operator and service with %d tries", maxDeleteLoops)
+		}
+		if i > 0 {
+			<-a.clock.After(3 * time.Second)
```

Comment:
needs to be in a select with a.catacomb.Dying()

Reply:
Done